server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3314,9 +3337,8 @@ assert getActiveOperationsCount() == OPERATIONS_BLOCKED
         SetOnce<Engine> newEngineReference = new SetOnce<>();
         final long globalCheckpoint = getLastKnownGlobalCheckpoint();
         assert globalCheckpoint == getLastSyncedGlobalCheckpoint();
-        synchronized (mutex) {
-            verifyNotClosed();
-            // we must create both new read-only engine and new read-write engine under mutex to ensure snapshotStoreMetadata,
+        synchronized (engineMutex) {
+            // we must create both new read-only engine and new read-write engine under engineMutex to ensure snapshotStoreMetadata,
             // acquireXXXCommit and close works.
             final Engine readOnlyEngine =
```

Comment:
I think we could have a race condition to `IndexShard.close()` here? If the shard is closed, store ref count may go to 0 and this would fail with `IllegalStateException` due to store.incRef(). This could get logged as an uncaught exception I think (which would fail tests)?

Reply:
I think it's fine but I might miss something here. If we hit an exception, we will fail this shard.

Reply:
OK, I guess the likelihood is very small here. I think failing the shards could print a warning only.